Theme editing feature

canstudios-nicolaw opened the pull request:
Please see feature information on https://github.com/adaptlearning/adapt_authoring/issues/2187

With these changes editable themes can expose less variables to the authoring tool. The values of these variables can be set from the theme editor page for a course. Edited less variables are saved to a file called "yyyyy.less" in the same fashion as custom less is currently saved to "zzzzz.less".

The "yyyyy.less" is included in export and these values are also applied to imported courses. If a preset has been selected then the preset will not be imported but the values will be.

If a course has an edited theme, and custom less, then the base theme styles should be applied, followed by the edited theme variables and lastly the custom less.

-------------------------------------------------------------

I have made a theme "Editable Vanilla" to be used for testing this feature. We would like to update the existing vanilla theme to make it editable (this theme is only to be used for testing purposes until the PR is out in vanilla). I've included the zip file below.

[adapt-contrib-editablevanilla.zip](https://github.com/adaptlearning/adapt_authoring/files/2822394/adapt-contrib-editablevanilla.zip)

-------------------------------------------------------------

Our QA team have had a look at this, and there are two small issues. As I am away next week I'm going to make this PR now and describe the issues I am aware of, I will fix these as soon as I am back.

1) If you select a preset on a theme, and then change the theme and come back, the preset will be selected in the dropdown when it shouldn't be (the values will be the default no preset values).
2) Selecting asset, then toggle preset results in default asset thumbnail. Clicking save after this and previewing the course will result in the old asset appearing, despite saving the themes preset.

-------------------------------------------------------------

If anybody has time to look at this while I'm away it would be much appreciated, hopefully we can get this long-awaited feature out soon!

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -728,9 +516,194 @@ OutputPlugin.prototype.applyTheme = function(tenantId, courseId, jsonObject, des
+            savedSettings[prop] = "\"" + assetPathArray.join('/').replace('course/', '') + "\"";
+          }
+
+          modifiedProperties += '@' + props[prop] + ' : ' + savedSettings[prop] + ';\n';
```
```suggestion
          modifiedProperties += '@' + props[prop] + ': ' + savedSettings[prop] + ';\n';
```

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -728,9 +516,194 @@ OutputPlugin.prototype.applyTheme = function(tenantId, courseId, jsonObject, des
+  ], function(error) {
+    if (error) {
+      logger.log('error', error);
+      return next(error);
```
Suggest moving this to replace line 703 – only one `return` needed.

tomgreenfield commented on plugins/content/config/model.schema:
```diff
@@ -291,6 +291,10 @@
     "themeSettings": {
       "type": "object"
+    },
+    "_themePreset": {
```
Note: needs adding to framework.

canstudios-nicolaw replied:
I've added a separate issue to this milestone, otherwise this is going to get lost https://github.com/adaptlearning/adapt_authoring/issues/2220

tomgreenfield commented on plugins/content/course/model.schema:
```diff
@@ -590,6 +590,9 @@
     "themeSettings": {
       "type": "object"
     },
+    "themeVariables": {
```
Note: needs adding to framework.

canstudios-nicolaw replied:
I've added a separate issue to this milestone, otherwise this is going to get lost https://github.com/adaptlearning/adapt_authoring/issues/2220

tomgreenfield commented on plugins/content/themepreset/index.js:
```diff
@@ -0,0 +1,99 @@
+  // save to config
+  app.contentmanager.retrieve('config', { _courseId: courseId }, function(err, config) {
+    if (err) return next(err);
+      var delta = config[0];
```
Please sort the indentation in this file.

canstudios-nicolaw replied:
Sorted now

tomgreenfield commented on plugins/output/adapt/export.js:
```diff
@@ -76,6 +78,49 @@ function copyFrameworkFiles(results, filesCopied) {
+    if (error) {
+      return variablesWritten(error);
+    }
+    var themeVariables = data['course'][0].hasOwnProperty('themeVariables')
```
These variables do not need to be checked for existence; they can be falsies:
```js
const themeVariables = data.course[0].themeVariables;
```

tomgreenfield commented on plugins/output/adapt/export.js:
```diff
@@ -76,6 +78,49 @@ function copyFrameworkFiles(results, filesCopied) {
+
+    const destinationFolder = path.join(EXPORT_DIR, 'src', 'theme', themeName);
+
+    if (themeVariables) {
```
Suggest reversing the logic to return early.

tomgreenfield commented on plugins/output/adapt/export.js:
```diff
@@ -76,6 +78,49 @@ function copyFrameworkFiles(results, filesCopied) {
+          var theme = results[0];
+
+          self.writeThemeVariables(COURSE_ID, theme, themeVariables, destinationFolder, function(error) {
+            if (error) {
```
Don't need the anonymous function here – can simply pass in the callback `variablesWritten`.

tomgreenfield commented on routes/lang/en-application.json:
```diff
@@ -1,4 +1,38 @@
 {
+  "app.themesettings": "Theme settings",
```
I don't believe this string is used anywhere.

tomgreenfield commented on routes/lang/en-application.json:
```diff
@@ -1,4 +1,38 @@
+  "app.preseteditdesc": "You can view, change name and delete the saved presets for the chosen theme here.",
+  "app.presetdeletetext": "This will delete the saved '<b>%{preset}</b>' preset. <br/><br/><b>IMPORTANT:</b> <em>This will affect any existing courses using this preset.<em>",
+  "app.presetnameplaceholder": "Enter new name",
+  "app.themecustomisation": "Theme customisation",
```
Is this string being used anywhere?

tomgreenfield commented on routes/lang/en-application.json:
```diff
@@ -1,4 +1,38 @@
+  "app.themingtitle": "Set the style for your course",
+  "app.themingdesc": "You can modify the appearance of your course on this page.",
+  "app.choosethemetitle": "Choose theme",
+  "app.choosethemedesc": "Select the base theme for your course and apply any saved 'preset' styles using the two dropdowns below.<br/><br/><em><b>IMPORTANT:</b> Changing the options here will replace any unsaved selections you've made in the form below; make sure there's nothing you want to keep!</em>",
```
Please remove all `<em>`s from these strings – we don't embed the italic version of Open Sans so we get undesirable obliques.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,29 +448,23 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
-  var themeAssetsFolder = path.join(destinationFolder, Constants.Folders.Assets);
-  var themeSettings = jsonObject['course'][0].hasOwnProperty('themeSettings')
-    ? jsonObject['course'][0].themeSettings
+  var themeVariables = jsonObject['course'][0].hasOwnProperty('themeVariables')
```
These variables do not need to be checked for existence; they can be falsies:
```js
var themeVariables = jsonObject.course[0].themeVariables;
```

taylortom commented:
Many thanks for tidying this up Nicola, it's looking really good.

A few points:
- With regards to the customisations of your editable theme, I'm not sure what should have changed on preview, so could you please give a very quick summary of what each var does? In particular, I don't seem to be seeing:
  - Tertiary
  - Quaternary 
  - Header font colour 
  - Item font color (typo in colour here)

- Can we get some styling for the preset dropdown when disabled? Currently it looks clickable
- It'd be nice if we could style the page to better match the other editor pages by separating out the nested form items into separate container 'blocks' (see screenshot below). Not sure how much work would be involved in this (as a start, I think we can remove the container style on the `.theme-customiser`, as the fieldsets have this styling already).

I can ask around to see if we have any themes with a different look & feel to vanilla that we can repurpose and release alongside this feature to give the community some variety.

![screenshot 2019-02-08 at 13 53 03](https://user-images.githubusercontent.com/1059083/52482641-bbb24280-2ba9-11e9-9088-1bf57b0f3a97.png)

canstudios-nicolaw commented:
Thanks for all your feedback @tomgreenfield and @taylortom :)

canstudios-nicolaw commented:
I'm making changes to this feature branch now, I will comment when it is ready for review.

canstudios-nicolaw commented:
@taylortom Regarding the editable theme, I think it would be best if it had it's own repo (which will start as a copy of vanilla) and then I can make a PR to that so that these comments don't get mixed in with the feature comments. If I do this, would we be happy to go with "adapt-contrib-editableVanilla" as the theme name?

Good point about the styling of different blocks. I will look into that.

canstudios-nicolaw commented:
@taylortom the more I think about styling different blocks the less certain I am. I understand your point about making the UI as consistent as possible, and I agree that is important. At the moment all of the settings are contained within the "Customise theme" block, and I think that makes it clear that the "Save as preset" button applies to them all. If I put them in their own blocks, I'm not sure if the "Customise theme" block would make sense on it's own. I could remove the blue outlines from the groups of settings? That could make it feel more like other parts of the UI?

taylortom commented:
@canstudios-nicolaw cool, I get your point.

One of the main reasons I mention the styling is I'd like to see us add support for the sidebar filters like we have for the other content. Think this could be particularly useful for theming (could group colour settings/paddings etc. etc.). Were we to do this, then we may have to move the save preset button as you mention. If there's no appetite for this, then your suggestion to remove the borders would be ideal.

canstudios-nicolaw commented:
@taylortom It's a good point. We will have a think about this internally as well. I'll be opening this up for re-review soon and we can see what the general consensus is :)

tomgreenfield commented on plugins/content/theme/index.js:
```diff
@@ -4,23 +4,23 @@
+const rest = require('../../../lib/rest');
+const usermanager = require('../../../lib/usermanager');
+const util = require('util');
+const _ = require('underscore');
```
Not sure that we need Underscore declared here?

canstudios-nicolaw replied:
You're right (there were a few others as well, I've removed them)

tomgreenfield commented on plugins/content/theme/index.js:
```diff
@@ -4,23 +4,23 @@
+const ContentTypeError = contentmanager.errors.ContentTypeError;
+const database = require('../../../lib/database');
+const defaultOptions = require('./defaults.json');
+const fs = require('fs');
```
```suggestion
const fs = require('fs-extra');
```

canstudios-nicolaw replied:
This wasn't used so was removed

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,244 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
 OutputPlugin.prototype.applyTheme = function(tenantId, courseId, jsonObject, destinationFolder, next) {
   var self = this;
+  var themeVariables = jsonObject.course[0].themeVariables;
+  var themeName = jsonObject['config'][0]._theme;
```
Suggest mirroring the line above:
```suggestion
  var themeName = jsonObject.config[0]._theme;
```

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,244 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+        db.retrieve('courseasset', {_courseId: courseId, _contentType: 'theme'}, function(err, results) {
+          if (err) {
+              seriesCallback(err);
+          } else if (results) {
```
Suggest returning early if no assets to process.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,244 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+      // This is in the case where an object has been used to group
+      // theme properties
+      async.eachSeries(_.keys(theme.properties), function(key, innerCallback) {
+        if (theme.properties[key].hasOwnProperty('properties')) {
```
Good candidate for caching and reversing the logic to return early?
```js
var nestedProps = theme.properties[key].properties;

if (!nestedProps) {
```

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,546 @@
+    loadCollections: function() {
+      this.themes = new ThemeCollection();
+      this.listenTo(this.themes, {
+        'sync': this.onCollectionReady,
```
No need for quotes on keys here.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,546 @@
+
+      this.presets = new PresetCollection();
+      this.listenTo(this.presets, {
+        'sync': this.onCollectionReady,
```
No need for quotes on keys here.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,546 @@
+        type: 'warning',
+        text: Origin.l10n.t('app.restorepresettext'),
+        callback: function(confirmed) {
+          if (confirmed) {
```
Suggest returning early.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,546 @@
+      // add 'no presets'
+      select.append($('<option>', { value: "", disabled: 'disabled', selected: 'selected' }).text(Origin.l10n.t('app.selectinstr')));
+      // add options
+      _.each(this.themes.models, function(item, index) {
```
Not sure whether we can be bothered to update but any `_.each()` loops on arrays can use the native `.forEach()`. 

Similarily, `_.keys()` → `Object.keys()`.

We only ever used the Underscore versions in Adapt for IE8(!).

canstudios-nicolaw replied:
I've changed the ones in the themeEditor directory

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorPresetEditView.js:
```diff
@@ -0,0 +1,82 @@
+      Origin.Notify.confirm({
+        text: Origin.l10n.t('app.presetdeletetext', { preset: presetName }),
+        callback: function(confirmed) {
+          if (confirmed === true) {
```
Reverse logic?
```js
if (!confirmed) return;
```

tomgreenfield commented on frontend/src/modules/editor/themeEditor/models/editorModel.js:
```diff
@@ -0,0 +1,125 @@
+      // Remove any attributes which are not on the whitelist
+      // Useful to call before a save()
+      pruneAttributes: function() {
+        var self = this;
```
Suggest:
```js
if (!this.whitelistAttributes) return;

// Ensure that only valid attributes are pushed back on the save
Object.keys(this.attributes).forEach(function(key) {
  if (!_.contains(this.whitelistAttributes, key)) {
    this.unset(key);
  }
}, this);
```

lc-thomasberger commented on frontend/src/modules/editor/themeEditor/templates/editorPresetEdit.hbs:
```diff
@@ -0,0 +1,30 @@
+          <a href="" class="button cancel btn primary-hollow">
+            {{t 'app.themepresetcancel'}}
+          </a>
+        </div>
```
We should not use `<a>` for buttons. Can you please update to use `<button>`.

lc-thomasberger commented:
Thanks for this PR and providing the test theme! I like the feature.

![image](https://user-images.githubusercontent.com/11990374/52704001-e58bb080-2f7f-11e9-9828-9472fb355a61.png)

I think the "manage presets" modal is a bit to big. This also makes the alignment of the preset-items look a bit odd.
I would suggest adding a max-width in px. The items should use the whole width of the modal. This can easily done using flexbox.

The close button of the modal is not well aligned. Suggest you use position absolute instead of float right.

When you click "Edit name" it is hard to see that the name is actually editable. I think it would be good and easy to set the focus to the input once you click the "Edit name" button.

In general, I think we should try to avoid `<a href="#">` for buttons. Also we should try to use flexbox instead of floats. Flexbox is far more powerful and well supported.

taylortom commented:
The bad styling is probably on me, so apologies in advance for that 😅

lc-thomasberger commented:
I think there is an issue with updating presets.
If you update a colour of an existing preset and hit save and revisit the theme editor, everything is fine.
If you select no-preset and then the updated preset, it displays the initial settings. So the updated values are lost.

canstudios-nicolaw commented:
Thank you for the review @lc-thomasberger all good points about the modal styling :) 

I'm not sure that the issue with updating presets sounds like an issue to me. I would expect that if you change presets in the dropdown you should see the default values for that preset. In my mind when some settings are saved they are just saved as individual settings, they are not saved as 'this preset with these settings on top', so if I changed presets and came back I wouldn't expect them to be remembered as part of the preset. I would also imagine that if you want to change some settings on a preset and then keep them, then that should be a new preset. Alternatively, if you were to click 'Cancel' in the scenario you describe then I think that the updated values would not have been lost

@taylortom no worries :joy: I should have spotted it before PR!

lc-thomasberger commented:
@canstudios-nicolaw valid point. Altough I think there is a use case for updating presets. Maybe we can add a update preset button once a preset is changed? Can be a separate issue though if it requires to much refactoring.

canstudios-nicolaw commented:
@lc-thomasberger I don't think updating presets is MVP for this feature as it's already possible to create and delete presets. I will raise an issue and add it to this milestone for now so that it isn't forgotten - I do think it would be nice to have.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,547 @@
+      // add 'no presets'
+      select.append($('<option>', { value: "", disabled: 'disabled', selected: 'selected' }).text(Origin.l10n.t('app.selectinstr')));
+      // add options
+        this.themes.models.forEach(function(item) {
```
Please fix indentation.

tomgreenfield replied:
Sorry, this is still off; lines 141 & 142 need to be less indented.

canstudios-nicolaw replied:
For some reason, this looks absolutely fine in my IDE :( I'll have to have a play around with it

tomgreenfield replied:
```suggestion
      this.themes.models.forEach(function(item) {
```

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,547 @@
+    },
+
+    postSettingsData: function(callback) {
+      if (this.form) {
```
Suggest reversing logic to return early.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+              }
+            });
+          } else {
+              savedSettings[key] = themeVariables[key];
```
Please fix indentation.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+                  outputFilename = path.join(themeAssetsFolder, asset.filename);
+
+              // Ensure that an asset is only copied once
+              if (!processedAssets[asset.filename]) {
```
Suggest switching logic to return early.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+      // Flatten the property names to allow two levels
+      // This is in the case where an object has been used to group
+      // theme properties
+      async.eachSeries(_.keys(theme.properties), function(key, innerCallback) {
```
Please switch to native `Object.keys()`.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
-    ? jsonObject['config'][0]._theme
-    : false;
+        // There are nested properties to process
+        async.eachSeries(_.keys(nestedProps), function(childKey, secondInnerCallback) {
```
Please switch to native `Object.keys()`.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+    },
+    function(seriesCallback) {
+      // Now flatten the themeVariables
+      async.eachSeries(_.keys(themeVariables), function(key, innerCallback) {
```
Please switch to native `Object.keys()`.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+        if (key !== '_type') {
+          if (typeof themeVariables[key] === 'object') {
+            // Iterate the properties and add them to the array
+            async.each(_.keys(themeVariables[key]), function(childKey, secondInnerCallback) {
```
Please switch to native `Object.keys()`.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,547 @@
+      var selectedTheme = this.getSelectedTheme();
+      var selectedThemeId = selectedTheme.get('_id');
+      $.post('/api/theme/' + selectedThemeId + '/makeitso/' + this.model.get('_courseId'))
+        .error(_.bind(this.onSaveError, this))
```
Please could you switch to native `.bind()` in this file?

tomgreenfield commented on frontend/src/modules/editor/themeEditor/less/editorPresetEdit.less:
```diff
@@ -0,0 +1,76 @@
+        }
+
+        .button {
+          padding-right: 10px;
```
Not sure what this styling does apart from make the text slightly off-centre?

tomgreenfield commented on frontend/src/modules/editor/themeEditor/less/editorPresetEdit.less:
```diff
@@ -0,0 +1,76 @@
+
+    .button {
+        display: inline-block;
+        width: 115px;
```
Needs a `vertical-align: middle;` to stop the buttons going out of alignment.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/less/editorPresetEdit.less:
```diff
@@ -0,0 +1,76 @@
+          display: inline-block;
+          width: 50%;
+          border: 1px solid #f1f1f1;
+          border-radius: 5px;
```
Needs a `vertical-align: middle;` to align with the buttons.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,546 @@
+
+    postSettingsData: function(callback) {
+      if (!this.form) {
+        callback.apply(this);
```
Missing a `return`.

canstudios-nicolaw commented:
I have now made PRs to update the vanilla theme, please could these themes be used for testing and any comments made on the relevant PR https://github.com/adaptlearning/adapt_framework/issues/2359

canstudios-nicolaw commented:
@tomgreenfield Happy to move the buttons if that is the preference. I had a look around the UI of the authoring tool and I couldn't find anywhere that had save buttons in the main body of the page, I also thought that a benefit of putting them in the sidebar would mean that users wouldn't have to scroll back to the top to save the preset.

@taylortom @lc-thomasberger Please could you let me know your opinions on button placement?

I will look into tweaking the code as you suggest TomG

lc-thomasberger commented:
I do like the rows suggested by Tom G' Screenshot.
Manage Presets makes sense to be next to the preset dropdown.

"Save as Preset" and "Restore to Preset" is related to the variables and therefore I think it should be part of the variable group below.

![image](https://user-images.githubusercontent.com/11990374/53330375-95e9a500-38ee-11e9-8a54-9e095245220f.png)

canstudios-nicolaw commented:
@lc-thomasberger I don't think the restore and save buttons can be with the variables anymore. Variables could be split across several separate boxes now (eg Colours, Fonts etc) so it would be strange if these buttons only existed for the first box but applied to variables across all the boxes.

lc-thomasberger commented:
Then maybe we leave them in the sidebar for now ... 😄

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -442,295 +448,248 @@ OutputPlugin.prototype.clearBuildFlag = function(pathToBuildFlag, next) {
+
+        if (!nestedProps) {
+          props[key] = key;
+          innerCallback();
```
Missing a `return`.

tomgreenfield commented on routes/lang/en-application.json:
```diff
@@ -1,4 +1,36 @@
+  "app.themeeditor": "Theming",
+  "app.savepreset": "Save as preset",
+  "app.themingtitle": "Set the style for your course",
+  "app.themingdesc": "You can modify the appearance of your course on this page.",
```
Is this string is used anywhere?

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,549 @@
+      $('option', select).remove();
+      // add 'no presets'
+      select.append($('<option>', { value: "", disabled: 'disabled', selected: 'selected' }).text(Origin.l10n.t('app.selectinstr')));
+        // add options
```
```suggestion
      // add options
```

tomgreenfield commented on plugins/content/config/model.schema:
```diff
@@ -291,6 +291,10 @@
     "themeSettings": {
       "type": "object"
+    },
+    "_themePreset": {
```
With this attribute I'm getting the classic error on the configuration settings page:
```console
Uncaught TypeError: constructorFn is not a constructor
    at child.createEditor (backbone-forms.js:842)
    at child.initialize (backbone-forms.js:803)
    at child.Backbone.View (backbone.js:991)
    at new child (backbone.js:1541)
    at child.createField (backbone-forms.js:136)
    at child.<anonymous> (backbone-forms.js:83)
    at Array.forEach (<anonymous>)
    at Function._.each._.forEach (underscore.js:79)
    at child.initialize (backbone-forms.js:81)
    at child.Backbone.View (backbone.js:991)
```

canstudios-nicolaw replied:
Were you using the correct framework branch? https://github.com/adaptlearning/adapt_framework/issues/2360

tomgreenfield replied:
Unrelated – this branch currently utilises the core schema which is where it's picking up this attribute.

I would've thought it just needs adding to the [blacklist](https://github.com/CanStudios/adapt_authoring/blob/48d005e64a57627e9beff44b2911487321b839de/lib/contentmanager.js#L561-L586).

canstudios-nicolaw commented:
@adaptlearning/authoring-tool-push Could I have a hand?

I have rebased the feature/theme-editing branch in the canstudios fork onto the latest master. I have also rebased release/theme-editing onto the latest master.

If I compare these branches as if I was going to make a new PR from the canstudios fork I can see the correct number of commits (just the theme editing changes): https://github.com/adaptlearning/adapt_authoring/compare/release/theme-editing...CanStudios:feature/theme-editing

However this PR is showing all the commits for 0.7.0. 

Is there something I need to do to 'refresh' this PR on github?

taylortom commented:
@canstudios-nicolaw just a wild stab in the dark 😅

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -77,18 +77,17 @@ var Constants = {
       Package: 'package.json',
       Metadata: 'metadata.json',
-      ThemeVariables: 'yyyyy.less',
+      ThemeVariables: '1-themeVariables.less',
```
I don't think `Constants.Filenames.Variables` is used any longer so could switch to that instead of having `ThemeVariables`.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -77,18 +77,17 @@ var Constants = {
       Main: 'index.html',
       Rebuild: '.rebuild',
-      CustomStyle: 'zzzzz.less',
+      CustomStyle: '2-customLess.less',
```
Suggest `2-customStyles.less`.

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -606,12 +599,18 @@ OutputPlugin.prototype.writeThemeVariables = function(courseId, theme, themeVari
           logger.log('error', 'Theme customisations 4 of 4');
         }
-        seriesCallback(err);
+        seriesCallback(null);
```
Did you mean to change from `err` to `null`? Looks like any error would now be ignored.

tomgreenfield commented:
Hi @canstudios-nicolaw, this still isn't working for me I'm afraid – upon loading of the editor it looks as though the code is comparing against the default (no preset) settings rather than those of the current preset.

canstudios-nicolaw commented:
@tomgreenfield I think perhaps I fixed a slightly different issue then (oops). The one I fixed was that for themes with nested properties, the 'Restore to preset' button wasn't shown if I made changes from the preset, saved and then returned to the page.

So, just to confirm, the issue that you're seeing happens when:
- Select a theme with unnested properties
- Select a preset
- Make some changes on top of that preset
- Save
- Go back to the theme editor
- 'Save to Preset' and 'Restore to preset' buttons not visible

Is that correct?

(Sorry I've just switched the testing server over to bugpatch 1 for QA and I'm in the middle of something right now so don't have an easy way to look at this)

canstudios-nicolaw commented:
Ok brilliant thanks, I will have a look at this tomorrow.

canstudios-nicolaw commented:
@tomgreenfield I've fixed this for the restore to preset button. At the moment the 'Save to preset' button is always visible by design. Personally I think I'm in favour of keeping it that way. I think it makes sense that you can only restore to preset when you have made changes, but I don't think there is a need to restrict when you can create one.

Worst case scenario you don't realise you already have a preset and end up with two identical ones, but it is possible to delete duplicates.

If you feel strongly that it shouldn't always be visible I can quite easily make it appear in the same fashion as the restore button I'm not sure if it could get confusing.

tomgreenfield commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,560 @@
+    },
+
+    onFieldChanged: function() {
+        this.updateRestorePresetButton(true);
```
I'm seeing that when you change a colour and then put it back to the preset's value, the restore to preset button is still showing when it shouldn't.

It should be as simple as removing the parameter from this function call but the code doesn't seem to be properly calculating the current selections?

Btw, completely agree with you re: the save button.

canstudios-nicolaw replied:
Yes you're right, the code seems to be comparing what is currently saved against the preset. It should be comparing what is currently selected in the form against the preset.

Looking in to how to get what is currently selected in the form. If you know how to to this, or you know of somewhere else in the authoring tool that might be doing something similar, then please let me know because it doesn't seem straightforward to me at the moment.

tomgreenfield replied:
Reverse-engineering from the way values are set in the current code, this sort of thing will work for unnested properties:
```js
getCurrentSettings: function() {
  if (!this.form) {
    return Origin.editor.data.course.get('themeVariables');
  }

  return _.mapObject(this.form.fields, function(field) {
    return field.getValue();
  });
},
```

The get functions (above and `getDefaultThemeSettings()`) should probably be modified to flatten any nested properties so that it's simpler to compare.

canstudios-nicolaw commented:
I have updated vanilla: https://github.com/adaptlearning/adapt-contrib-vanilla/pulls

tomgreenfield commented on lib/outputmanager.js:
```diff
@@ -234,6 +234,11 @@ OutputPlugin.prototype.sanitizeCourseJSON = function(mode, json, next) {
     }
 
+    // _themePreset should not be included in export
+    if (mode === Constants.Modes.Export) {
```
Question: why solely not in export? It's included in publishes at present.

canstudios-nicolaw replied:
I wish I could remember the answer, my fault for not documenting a reason at the time.. 

I will test removing it in publishes and hopefully remove from there also.

canstudios-nicolaw commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,576 @@
+    events: {
+      'change .theme select': 'onThemeChanged',
+      'change .preset select': 'onPresetChanged',
+      'change .form-container form': 'updateRestorePresetButton',
```
I'm fine with putting this back into the onFieldChanged handler, please could you explain how it's DRYer though? I just figured it made more sense to have the event call the function directly rather than call a function whose only purpose is to call another function?

tomgreenfield replied:
Sure. I envisioned the UI to be tweaked somehow to make it obvious we've customised our theme/preset, e.g.:
![modified](https://user-images.githubusercontent.com/922987/56348336-1ecde000-61be-11e9-8850-5827691db32e.png)
I nearly requested this in a review but I didn't want to burden the workload any more at this point! With the current code, any further UI changes would require repeating the event handler:
```js
'change .form-container form': 'updateRestorePresetButton',
'change .form-container form': 'updateSelectText'
```
The other undesired side effect of not having a dedicated event handler, as you've already noticed, is that we don't have control over the arguments. The logic had to be changed [here](https://github.com/adaptlearning/adapt_authoring/pull/2209/commits/97179a923c5cda712cc5e9d3dcdd1e94db4e7cd4#diff-18eb4043de18e1156fb7293f6fa554e1R453) because the `shouldShow` parameter is now sometimes the change event.

In an ideal world, there would be no comparisons being done in the UI functions – instead we'd derive `shouldShow`/`isThemeModified` in the event handler or even in its own function so it's reusable, and then pass a toggle to one/more UI setters.

Hope I'm making sense!

canstudios-nicolaw replied:
Ok fair point, I will put it back to how it was. Thanks for explaining.

I would have created an event handler function instead of having the same event calling multiple functions for sure but I see your point about the logic.

I like the idea of tweaking the UI in this way, happy to raise an issue for it and work on it (perhaps as part of a bug patch) but I agree we shouldn't be adding more to this PR/release at this point.

canstudios-nicolaw commented on frontend/src/modules/editor/themeEditor/views/editorThemingView.js:
```diff
@@ -0,0 +1,576 @@
+
+    getDefaultThemeSettings: function() {
+      var defaults = {};
+      var props = this.getSelectedTheme().attributes.properties;
```
I will look into this.

canstudios-nicolaw replied:
I think within the variables object it is still going to be important to specify a title for each group of variables (colours, fonts etc.), so I think it will have to be more like this:

```
"variables": {
      "_colors": {
        "type": "object",
        "required": false,
        "title": "Colours",
        "properties": {
          "primary-color": {
            "title": "Primary",
            "help": "This is the primary colour used to style the course, and is applied to component items and pop-up notifications.",
            "type": "string",
            "inputType": "ColourPicker",
            "default": "#078292"
          }
        }
      }
    }
```

tomgreenfield replied:
That's what I was going for; just cut a bit too many attributes out of my snippet! Basically the same structure at present (nested & unnested) with a parent `variables` property.

canstudios-nicolaw commented:
@tomgreenfield I've added support for the change to the schema as suggested. I've also updated the vanilla PRs here https://github.com/adaptlearning/adapt-contrib-vanilla/pulls

tomgreenfield commented:
Thanks @canstudios-nicolaw! I have passed this onto our QA team.
